File: javascript/diamond/diamond.js
export class Diamond {
    constructor() {
        this.letters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';
    }

    *intSeq(from, to, step) {
        for (let i = from; (step > 0 ? i <= to : i >= to); i += step) {
            yield i;
        }
    }

    makeLine(n, size) {
        const inside = ' '.repeat(2 * n - 1);
        const outside = ' '.repeat(size - n);
        return `${outside}${this.letters[n]}${inside}${this.letters[n]}${outside}`;
    }

    makeDiamond(letter) {
        const size = this.letters.indexOf(letter);
        if (size === 0) {
            return 'A\n';
        } else {
            const aLine = `${' '.repeat(size)}A${' '.repeat(size)}`;
            return `${aLine}
${[...this.intSeq(1, size, 1)].concat([...this.intSeq(size - 1, 1, -1)]).map((i) => this.makeLine(i, size)).join('\n')}
${aLine}
`;
        }
    }
}

File: clojure/diamond/src/diamond.clj
(ns diamond)

(defn- whitespace [n]
  (apply str (repeat n " ")))

(defn- limit-row [last-letter]
  (let [external-whitespace (whitespace (- (int last-letter) (int \A)))]
    (str external-whitespace "A" external-whitespace)))

(defn- row [last-letter letter]
  (if (= letter \A)
    (limit-row last-letter)
    (let [external-whitespace (whitespace (- (int last-letter) (int letter)))
          internal-whitespace (whitespace (dec (* 2 (- (int letter) (int \A)))))]
      (str external-whitespace letter internal-whitespace letter external-whitespace))))

(defn- rows [last-letter]
  (let [letters (map char (range (int \A) (inc (int last-letter))))]
    (map (partial row last-letter) letters)))

(defn diamond [last-letter]
  (let [forward-rows (rows last-letter)
        backward-rows (reverse (drop-last forward-rows))]
    (concat forward-rows backward-rows)))
